packages/frontend/core/src/pages/workspace/index.tsx

```diff
@@ -37,24 +38,104 @@ declare global {
 
 export const Component = (): ReactElement => {
   performanceRenderLogger.debug('WorkspaceLayout');
+  const { workspacesService } = useServices({
+    WorkspacesService,
+  });
 
   const params = useParams();
+  const location = useLocation();
 
-  const [showNotFound, setShowNotFound] = useState(false);
-  const workspacesService = useService(WorkspacesService);
-  const listLoading = useLiveData(workspacesService.list.isLoading$);
-  const workspaces = useLiveData(workspacesService.list.workspaces$);
+  // check if we are in detail doc route, if so, maybe render share page
+  const detailDocRoute = useMemo(() => {
+    const match = matchPath(
+      '/workspace/:workspaceId/:docId',
+      location.pathname
+    );
+    if (
+      match &&
+      match.params.docId &&
+      match.params.workspaceId &&
+      // TODO(eyhn): need a better way to check if it's a docId
+      viewRoutes.find(route => matchPath(route.path, '/' + match.params.docId))
+        ?.path === '/:pageId'
+    ) {
+      return {
+        docId: match.params.docId,
+        workspaceId: match.params.workspaceId,
+      };
+    } else {
+      return null;
+    }
+  }, [location.pathname]);
 
+  const [workspaceNotFound, setWorkspaceNotFound] = useState(false);
+  const listLoading = useLiveData(workspacesService.list.isRevalidating$);
+  const workspaces = useLiveData(workspacesService.list.workspaces$);
   const meta = useMemo(() => {
     return workspaces.find(({ id }) => id === params.workspaceId);
   }, [workspaces, params.workspaceId]);
 
-  const workspace = useWorkspace(meta);
-  const globalContext = useService(GlobalContextService).globalContext;
+  // if listLoading is false, we can show 404 page, otherwise we should show loading page.
+  useEffect(() => {
```

Comment:
I think this section could be moved into route loader

Reply:
I don't like using router-loader, I think it's hard to control.

And with this we can also automatically recover from the 404 page after the workspace list changes (maybe the network is restored).

Reply:
How about making the steps 1~6 into a state machine in a service? It is hard to maintain the logic into individual useEffects.

Reply:
I encountered some difficulties when I tried it, and it doesn't seem easy to write it as a state machine.

But you are right, since we will need a mobile version of the UI in the future, we also need to share the logic here. Maybe we need to extract them to one place.

Reply:
We can discuss a suitable solution later.